perfetto_trace/gpu_slice_data.cpp

```diff
@@ -0,0 +1,264 @@
+    auto        it = m_trace_processor->ExecuteQuery(query_str);
+    if (!it.Status().ok())
+    {
+        std::cerr << "No submission found in the perfetto trace during the dive trace event."
+                  << std::endl;
+        return args_set_data;
+    }
+
+    for (uint32_t rows = 0; it.Next(); rows++)
+    {
+        ArgsData arg;
+
+        int i = 0;
+        GetQueryValue(it, i++, arg.m_id, fields);
+        GetQueryValue(it, i++, arg.m_type, fields);
+        GetQueryValue(it, i++, arg.m_arg_set_id, fields);
+        GetQueryValue(it, i++, arg.m_flat_key, fields);
+        GetQueryValue(it, i++, arg.m_int_value, fields);
+        GetQueryValue(it, i++, arg.m_string_value, fields);
+        GetQueryValue(it, i++, arg.m_real_value, fields);
+        GetQueryValue(it, i++, arg.m_value_type, fields);
+        args_set_data.emplace_back(arg);
+    }
+
+    return args_set_data;
+}
+
+SurfaceData GpuSliceDataParser::ParseSurfaceData(int64_t submission_id)
+{
+    SurfaceData surface_data;
+
+    std::vector<std::string> fields = { "id",   "ts",         "dur",         "submission_id",
+                                        "name", "arg_set_id", "render_pass", "command_buffer" };
+    std::string              field_str = absl::StrJoin(fields, ",");
+
+    std::string query_str = absl::
+    StrFormat("SELECT %s FROM gpu_slice WHERE submission_id = %ld AND name=\"Surface\"",
+              field_str,
+              submission_id);
+    ptp::Iterator it = m_trace_processor->ExecuteQuery(query_str);
+    if (!it.Status().ok())
+    {
+        std::cerr << "Failed to query surface data for submission_id:" << submission_id
+                  << std::endl;
+        return surface_data;
+    }
+    for (uint32_t rows = 0; it.Next(); rows++)
+    {
+        int i = 0;
+        GetQueryValue(it, i++, surface_data.m_surface_id, fields);
+        GetQueryValue(it, i++, surface_data.m_ts, fields);
+        GetQueryValue(it, i++, surface_data.m_duration, fields);
+        GetQueryValue(it, i++, surface_data.m_submission_id, fields);
+        GetQueryValue(it, i++, surface_data.m_name, fields);
+        GetQueryValue(it, i++, surface_data.m_arg_set_id, fields);
+        GetQueryValue(it, i++, surface_data.m_render_pass, fields);
+        GetQueryValue(it, i++, surface_data.m_command_buffer, fields);
+        break;  // should at most have 1 row.
+    }
+    // Parse args from args table
+    surface_data.m_args = ParseArgsData(surface_data.m_arg_set_id);
+
+    return surface_data;
+}
+
+std::vector<SurfaceData> GpuSliceDataParser::ParseSurfaceData()
+{
+    std::vector<SurfaceData> surface_data;
+    std::vector<int64_t>     start_end_ts = ParseDiveTraceTimestamp();
+    assert(start_end_ts.size() == 2);
+
+    std::vector<int64_t> submission_ids = ParseSubmissionIds(start_end_ts[0], start_end_ts[1]);
+
+    for (auto submission_id : submission_ids)
+    {
+        surface_data.emplace_back(ParseSurfaceData(submission_id));
```

Comment:
could surface_data be large? if this field is large, maybe better with passing a & into the function (instead return a copy) and update directly to avoid an extra copy?

Reply:
oh, that is cool! good to know!